doc/xxhash_spec.md

```diff
@@ -324,6 +333,461 @@ The `XXH64()` function produces an unsigned 64-bit value as output.
+  u64 high = higherHalf(mulResult); // mulResult >> 64
+  u64 low = lowerHalf(mulResult); // mulResult & 0xFFFFFFFFFFFFFFFF
+  high = high + (low << 1);
+  low = low xor (high >> 3);
+  low = low xor (low >> 35);
+  low = low * PRIME_MX2;
+  low = low xor (low >> 28);
+  high = avalanche(high);
+  return {low, high};
+```
+
+### 9-16 bytes of input
+
+The algorithm starts from reading the first and last 8 bytes of the input as little-endian 64-bit values:
+
+```c
+u64 inputFirst = input[0:8];
+u64 inputLast = input[inputLength-8:inputLength];
+```
+
+Once again, these values are combined with a segment of the secret to produce the final value.
+
+```c
+XXH3_64_9to16():
+  u64 secretWords[4] = secret[24:56];
+  u64 low = ((secretWords[0] xor secretWords[1]) + seed) xor inputFirst;
+  u64 high = ((secretWords[2] xor secretWords[3]) - seed) xor inputLast;
+  u128 mulResult = (u128)low * (u128)high;
+  u64 value = inputLength + bswap64(low) + high + (u64)(lowerHalf(mulResult) xor higherHalf(mulResult));
+  return avalanche(value);
+
+XXH3_128_9to16():
+  u64 secretWords[4] = secret[32:64];
+  u64 val1 = ((secretWords[0] xor secretWords[1]) - seed) xor inputFirst xor inputLast;
+  u64 val2 = ((secretWords[2] xor secretWords[3]) + seed) xor inputLast;
+  u128 mulResult = (u128)val1 * (u128)PRIME64_1;
+  u64 low = lowerHalf(mulResult) + ((u64)(inputLength - 1) << 54);
+  u64 high = higherHalf(mulResult) + ((u64)higherHalf(val2) << 32) + (u64)lowerHalf(val2) * PRIME32_2;
+  // the above line can also be simplified to higherHalf(mulResult) + val2 + (u64)lowerHalf(val2) * (PRIME32_2 - 1);
+  low = low xor bswap64(high);
+  // the following three lines are in fact a 128x64 -> 128 multiplication ({low,high} = (u128){low,high} * PRIME64_2)
+  u128 mulResult2 = (u128)low * (u128)PRIME64_2;
+  low = lowerHalf(mulResult2);
+  high = higherHalf(mulResult2) + high * PRIME64_2;
+  return {avalanche(low), // lower half
+          avalanche(high)}; // higher half
+```
+
+
+XXH3 Algorithm Description (for medium inputs)
+-------------------------------------
+
+This algorithm is used for medium inputs (17-240 bytes of input). Its internal hash state is stored inside 1 (XXH3-64) or 2 (XXH3-128) "accumulators", each storing an unsigned 64-bit value.
+
+### Step 1. Initialize internal accumulators
+
+The accumulator(s) are initialized based on the input length.
+
+```c
+// For XXH3-64
+u64 acc = inputLength * PRIME64_1;
+
+// For XXH3-128
+u64 acc[2] = {inputLength * PRIME64_1, 0};
+```
+
+### Step 2. Process the input
+
+This step is further divided into two cases: one for 17-128 bytes of input, and one for 129-240 bytes of input.
+
+#### Mixing operation
+
+This step uses a mixing operation that mixes a 16-byte segment of data, a 16-byte segment of secret and the seed into a 64-bit value as a building block. This operation treat the segment of data and secret as little-endian 64-bit values.
+
+```c
+mixStep(u8 data[16], size secretOffset, u64 seed):
```

Comment:
nit: this interface possibly miss the `secret` pointer,
the following code would only make sense if `secret` was a global variable only.

Reply:
I realize that not providing `secret` as an input parameter is a good way to improve brevity, which is beneficial for readability.
I think I'll keep it this way,
I just want to make it clear that `secret` is not _necessarily_, it can also be set by the user.
There's already a paragraph about this point, which is probably clear enough.
Let's just ensure it.